+nosnoc/+discrete_time_problem/Gcs.m

```diff
@@ -0,0 +1,870 @@
+                        model.x_ref_val(:,ii),...
+                        p);
+                end
+                if ~isempty(model.u_ref_val)
+                    obj.f = obj.f + t_stage*dcs.f_lsq_u_fun(obj.w.u(ii),...
+                        model.u_ref_val(:,ii),...
+                        p);
+                end
+                if ~opts.cost_integration
+                    obj.f = obj.f + dcs.f_q_fun(x_ijk, z_ijk, lambda_ijk, ui, v_global, p);
+                end
+
+                % Clock Constraints
+                % TODO(@anton) HERE BE DRAGONS. This is by far the worst part of current nosnoc as it requires the discrete problem
+                %              to understand something about the time-freezing reformulation which is ugly.
+                if obj.opts.use_fesd && opts.equidistant_control_grid
+                    if opts.time_optimal_problem
+                        if opts.use_speed_of_time_variables
+                            obj.g.equidistant_control_grid(ii) = {[sum_h - opts.h;sot*sum_h - obj.w.T_final()/opts.N_stages]};
+                        else
+                            obj.g.equidistant_control_grid(ii) = {sum_h - obj.w.T_final()/opts.N_stages};
+                        end
+                    else
+                        if opts.relax_terminal_numerical_time
+                            % TODO(@armin) why is this allowed to be negative?
+                            obj.w.s_numerical_time(ii) = {{'s_numerical', 1}, -2*opts.h, 2*opts.h, opts.h/2};
+                            g_eq_grid = [sum_h - t_stage - obj.w.s_numerical_time(ii);
+                                -(sum_h - t_stage) - obj.w.s_numerical_time(ii)];
+                            obj.g.equidistant_control_grid(ii) = {g_eq_grid, -inf, 0};
+                            obj.f = obj.f + opts.rho_terminal_numerical_time*obj.w.s_numerical_time(ii);
+                        else
+                            obj.g.equidistant_control_grid(ii) = {t_stage-sum_h};
+                        end
+                    end
+                end
+            end
+
+            x_end = obj.w.x(opts.N_stages,opts.N_finite_elements(opts.N_stages),opts.n_s+rbp);
+            z_end = obj.w.z(opts.N_stages,opts.N_finite_elements(opts.N_stages),opts.n_s+rbp);
+            % Terminal cost
+            obj.f = obj.f + dcs.f_q_T_fun(x_end, z_end, v_global, p_global);
+
+            % Terminal_lsq_cost
+            if ~isempty(model.x_ref_end_val)
+                obj.f = obj.f + h0*opts.N_finite_elements(ii)*dcs.f_lsq_T_fun(x_end,...
+                    model.x_ref_end_val,...
+                    p_global);
+            end
+
+            % Terminal constraint
+            if opts.relax_terminal_constraint_homotopy
+                error("Currently unsupported")
+            end
+            g_terminal = dcs.g_terminal_fun(x_end, z_end, v_global, p_global);
+            switch opts.relax_terminal_constraint
```

Comment:
add TODO or even issue, 
to refactor this constrain relaxation as soon as the VDX relaxation is mature. 
But in general, i would still ask via options which group/type of constraints should be relaxed and how 
e.g. only time constraints, terminal constraints, path constraints, dynamics constraints (diff and algebaric) and have meaningful defaults. this flexiblity might be crucial for the really hard problems

Reply:
I believe this is #96?